src/java_bytecode/java_qualifiers.cpp

```diff
@@ -0,0 +1,154 @@
+// Copyright 2018 DiffBlue Limited. All Rights Reserved.
+
+/// \file
+/// Java-specific type qualifiers
+
+#include "java_qualifiers.h"
+#include <sstream>
+#include <util/make_unique.h>
+#include "expr2java.h"
+
+java_qualifierst::java_qualifierst(const namespacet &ns) : ns(ns)
+{
+}
+
+java_qualifierst &java_qualifierst::operator=(const java_qualifierst &other)
+{
+  INVARIANT(
+    &other.ns == &ns,
+    "Can only assign from a java_qualifierst using the same namespace");
+  annotations = other.annotations;
+  c_qualifierst::operator=(other);
+  return *this;
+}
+
+java_qualifierst &java_qualifierst::operator=(java_qualifierst &&other)
+{
+  INVARIANT(
+    &other.ns == &ns,
+    "Can only assign from a java_qualifierst using the same namespace");
+  annotations = std::move(other.annotations);
+  c_qualifierst::operator=(other);
+  return *this;
+}
+
+c_qualifierst &java_qualifierst::operator=(const c_qualifierst &c_other)
+{
+  auto other = dynamic_cast<const java_qualifierst *>(&c_other);
+  if(other != nullptr)
+    *this = *other;
+  else
+  {
+    annotations.clear();
+    c_qualifierst::operator=(c_other);
+  }
+  return *this;
+}
+
+std::unique_ptr<c_qualifierst> java_qualifierst::clone() const
+{
+  auto other = util_make_unique<java_qualifierst>(ns);
+  *other = *this;
+  return std::move(other);
+}
+
+const std::vector<java_annotationt> &java_qualifierst::get_annotations() const
+{
+  return annotations;
+}
+
+std::size_t java_qualifierst::count() const
+{
+  return c_qualifierst::count() + annotations.size();
+}
+
+void java_qualifierst::clear()
+{
+  c_qualifierst::clear();
+  annotations.clear();
+}
+
+void java_qualifierst::read(const typet &src)
+{
+  c_qualifierst::read(src);
+  auto &annotated_type = static_cast<const annotated_typet &>(src);
+  annotations = annotated_type.get_annotations();
+}
+
+void java_qualifierst::write(typet &src) const
+{
+  c_qualifierst::write(src);
+  auto &annotated_type = static_cast<annotated_typet &>(src);
+  annotated_type.get_annotations() = annotations;
+}
+
+c_qualifierst &java_qualifierst::operator+=(const c_qualifierst &c_other)
+{
+  c_qualifierst::operator+=(c_other);
+  auto other = dynamic_cast<const java_qualifierst *>(&c_other);
+  if(other != nullptr)
+  {
+    std::copy(
+      other->annotations.begin(),
+      other->annotations.end(),
+      std::back_inserter(annotations));
+  }
+  return *this;
+}
+
+bool java_qualifierst::operator==(const c_qualifierst &c_other) const
```

Comment:
Define `operator==` with parameter type `java_qualifierst`. No need for a dynamic cast.

Reply:
This is an override of the virtual function with this signature on the base class.
This is needed to prevent a `c_qualifierst` comparing equal to a `java_qualifierst` that has annotations and to consequently to make sure that `operator==` is commutative.